File: kotlin-eclipse-ui/src/org/jetbrains/kotlin/ui/editors/quickassist/KotlinQuickAssistProcessor.java
package org.jetbrains.kotlin.ui.editors.quickassist;

import java.util.List;

import org.eclipse.core.runtime.CoreException;
import org.eclipse.jdt.ui.text.java.IInvocationContext;
import org.eclipse.jdt.ui.text.java.IJavaCompletionProposal;
import org.eclipse.jdt.ui.text.java.IProblemLocation;
import org.eclipse.jdt.ui.text.java.IQuickAssistProcessor;

import com.google.common.collect.Lists;

public class KotlinQuickAssistProcessor implements IQuickAssistProcessor {

    @Override
    public boolean hasAssists(IInvocationContext context) throws CoreException {
        return getAssists(context, null).length > 0;
    }

    @Override
    public IJavaCompletionProposal[] getAssists(IInvocationContext context, IProblemLocation[] locations)
            throws CoreException {
        List<KotlinQuickAssistProposal> allProposals = getSingleKotlinQuickAssistProposals();
        
        List<IJavaCompletionProposal> applicableProposals = Lists.newArrayList();
        for (KotlinQuickAssistProposal proposal : allProposals) {
            if (proposal.isApplicable()) {
                applicableProposals.add(proposal);
            }
        }
        
        for (KotlinQuickAssistProposalsGenerator proposalGenerator: getKotlinQuickAssistProposalsGenerators()) {
            if (proposalGenerator.isApplicable()) {
                applicableProposals.addAll(proposalGenerator.getProposals());
            }
        }
        
        return applicableProposals.toArray(new IJavaCompletionProposal[applicableProposals.size()]);
    }
    
    private List<KotlinQuickAssistProposal> getSingleKotlinQuickAssistProposals() {
        List<KotlinQuickAssistProposal> proposals = Lists.newArrayList();
        
        proposals.add(new KotlinReplaceGetAssistProposal());
        proposals.add(new KotlinSpecifyTypeAssistProposal());
        
        return proposals;
    }
    
    
    private List<KotlinQuickAssistProposalsGenerator> getKotlinQuickAssistProposalsGenerators() {
        List<KotlinQuickAssistProposalsGenerator> proposalsGenerators = Lists.newArrayList();
        
        proposalsGenerators.add(new KotlinAutoImportProposalsGenerator());
        
        return proposalsGenerators;
    }
}

File: kotlin-eclipse-ui/src/org/jetbrains/kotlin/ui/editors/quickassist/KotlinQuickAssistProcessor.kt
package org.jetbrains.kotlin.ui.editors.quickassist

import org.eclipse.core.runtime.CoreException
import org.eclipse.jdt.ui.text.java.IInvocationContext
import org.eclipse.jdt.ui.text.java.IJavaCompletionProposal
import org.eclipse.jdt.ui.text.java.IProblemLocation
import org.eclipse.jdt.ui.text.java.IQuickAssistProcessor
import com.google.common.collect.Lists
import java.util.ArrayList

object KotlinQuickAssistProcessor : IQuickAssistProcessor {
    override public fun hasAssists(context: IInvocationContext) : Boolean = getAssists(context, null).isNotEmpty()
    
    override public fun getAssists(context: IInvocationContext?, locations: Array<IProblemLocation>?) : Array<IJavaCompletionProposal> {
        val allApplicableProposals = ArrayList<IJavaCompletionProposal>()
        
        getSingleKotlinQuickAssistProposals().filterTo(allApplicableProposals) { it.isApplicable() }
        
        return allApplicableProposals.toTypedArray()
    }
    
<<<<<<< HEAD
    private fun getSingleKotlinQuickAssistProposals() : List<KotlinQuickAssistProposal> {
        return listOf(
            KotlinReplaceGetAssistProposal(), 
            KotlinSpecifyTypeAssistProposal(),
            KotlinRemoveExplicitTypeAssistProposal(),
            KotlinImplementMethodsProposal(),
            KotlinConvertToExpressionBodyAssistProposal(),
            KotlinConvertToBlockBodyAssistProposal(),
            KotlinChangeReturnTypeProposal())
=======
    private fun getSingleKotlinQuickAssistProposals(editor: KotlinEditor) : List<KotlinQuickAssistProposal> {
        return listOf(
            KotlinReplaceGetAssistProposal(editor), 
            KotlinSpecifyTypeAssistProposal(editor),
            KotlinRemoveExplicitTypeAssistProposal(editor),
            KotlinImplementMethodsProposal(editor),
            KotlinConvertToExpressionBodyAssistProposal(editor),
            KotlinConvertToBlockBodyAssistProposal(editor),
            KotlinChangeReturnTypeProposal(editor))
>>>>>>> abf1e11... Propagate editor through assists hierarchy to fix issues about uninitialized editor
    }
}